sdk/rust/src/lib.rs

```diff
@@ -0,0 +1,499 @@
+                .map_err(|_| serde::de::Error::custom(error)),
+        }
+    }
+}
+
+type WsRequest = (
+    api_types::ws::ClientRequest,
+    oneshot::Sender<ServerResultMessage>,
+);
+
+pub struct WsClientInner {
+    #[allow(dead_code)]
+    ws:             tokio::task::JoinHandle<()>,
+    request_sender: mpsc::UnboundedSender<WsRequest>,
+    request_id:     RwLock<usize>,
+
+    update_receiver: broadcast::Receiver<api_types::ws::ServerUpdateResponse>,
+}
+
+#[derive(Clone)]
+pub struct WsClient {
+    inner: Arc<WsClientInner>,
+}
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+enum MessageType {
+    Response(api_types::ws::ServerResultResponse),
+    Update(api_types::ws::ServerUpdateResponse),
+}
+
+pub struct WsClientUpdateStream<'a> {
+    stream:    BroadcastStream<api_types::ws::ServerUpdateResponse>,
+    _lifetime: PhantomData<&'a ()>,
+}
+
+impl WsClientUpdateStream<'_> {
+    pub fn new(stream: BroadcastStream<api_types::ws::ServerUpdateResponse>) -> Self {
+        Self {
+            stream,
+            _lifetime: PhantomData,
+        }
+    }
+}
+
+// Implementing Stream trait
+impl Stream for WsClientUpdateStream<'_> {
+    type Item = Result<api_types::ws::ServerUpdateResponse, BroadcastStreamRecvError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let stream = &mut self.get_mut().stream;
+        stream.poll_next_unpin(cx)
+    }
+}
+
+impl WsClient {
+    pub fn get_update_stream(&self) -> WsClientUpdateStream {
+        WsClientUpdateStream::new(BroadcastStream::new(
+            self.inner.update_receiver.resubscribe(),
+        ))
+    }
+
+    async fn run(
+        mut ws_stream: WebSocketStream<MaybeTlsStream<TcpStream>>,
+        mut request_receiver: mpsc::UnboundedReceiver<WsRequest>,
+        update_sender: broadcast::Sender<api_types::ws::ServerUpdateResponse>,
+    ) {
+        let mut ping_interval = tokio::time::interval(Duration::from_secs(30));
+        let mut responded_to_ping = true;
+        let mut requests_map = BTreeMap::<String, oneshot::Sender<ServerResultMessage>>::new();
+        loop {
+            tokio::select! {
+                message = ws_stream.next() => {
+                    let message = match message {
+                        Some(message) => {
+                            match message {
+                                Ok(message) => message,
+                                Err(_) => continue,
+                            }
+                        }
+                        None => break,
+                    };
+
+                    let message = match message {
+                        Message::Text(text) => {
+                            let response: Result<MessageType, serde_json::Error> = serde_json::from_str(&text);
+                            match response {
+                                Ok(response) => response,
+                                Err(_) => continue,
```

Comment:
shouldn't we log these?

Reply:
I think it's a good idea to add logging to rust client in a separate task maybe :-?